File: runner/progress.py
"""Live progress reporting for long batch runs.

A balance batch routinely runs tens of thousands of simulations, which for
minutes at a time looks indistinguishable from a hang. This renders a single
self-overwriting status line -- bar, percent, completed/total, throughput,
elapsed, and estimated time remaining -- while the batch streams results.

Two deliberate constraints:

* The line goes to **stderr**, so `batch > report.txt` and any piping of the
  report on stdout stay byte-for-byte clean.
* The reporter is a pure pass-through over the RunResult stream: it yields
  exactly the results it receives, in order, and never touches simulation
  state or the RNG. Turning progress on or off cannot change a batch's
  outcome for a given seed.

Throughput and ETA are computed from the overall average rate rather than an
instantaneous one. Results arrive from the process pool in bursts (one
dispatch window at a time), so a short-window rate would swing wildly while
the average converges quickly and stays readable.

Counting happens whether or not the line is being drawn, so callers can
report final timing stats even when output is not a terminal.
"""

import math
import shutil
import sys
import time

BAR_WIDTH = 28
MIN_REDRAW_SECONDS = 0.1

_BLOCKS = ("█", "░")  # full block, light shade
_ASCII_BLOCKS = ("#", "-")


def format_duration(seconds) -> str:
    """Render a duration as MM:SS, or H:MM:SS past an hour."""
    if seconds is None or seconds != seconds or seconds in (float("inf"), float("-inf")):
        return "--:--"
    total = max(0, int(round(seconds)))
    hours, remainder = divmod(total, 3600)
    minutes, secs = divmod(remainder, 60)
    if hours:
        return f"{hours:d}:{minutes:02d}:{secs:02d}"
    return f"{minutes:02d}:{secs:02d}"


def format_rate(rate) -> str:
    """Render simulations/second, keeping precision only where it informs."""
    if rate is None or rate != rate or rate in (float("inf"), float("-inf")):
        return "--"
    if rate >= 100:
        return f"{rate:,.0f}"
    if rate >= 10:
        return f"{rate:.1f}"
    return f"{rate:.2f}"


def render_line(completed, total, elapsed, width=None, bar_width=BAR_WIDTH, ascii_only=False):
    """Build the status line for a given point in a batch.

    Kept free of I/O and of wall-clock reads so the exact rendering can be
    asserted in tests from plain numbers.
    """
    fraction = 1.0 if total <= 0 else min(max(completed / total, 0.0), 1.0)
    rate = completed / elapsed if elapsed > 0 and completed > 0 else None
    remaining = (total - completed) / rate if rate else None
    if completed >= total:
        remaining = 0.0

    filled, empty = _ASCII_BLOCKS if ascii_only else _BLOCKS
    bar_width = max(1, bar_width)
    filled_cells = int(fraction * bar_width)
    bar = filled * filled_cells + empty * (bar_width - filled_cells)

    total_text = f"{total:,}"
    counts = f"{completed:,}".rjust(len(total_text)) + f"/{total_text}"

    # Floored, not rounded: 9,999 of 10,000 runs must not read as 100%.
    percent = math.floor(fraction * 1000) / 10

    line = (
        f"[{bar}] {percent:5.1f}% | {counts} | "
        f"{format_rate(rate)} sim/s | {format_duration(elapsed)} elapsed | "
        f"{format_duration(remaining)} left"
    )
    if width and len(line) > width:
        line = line[:width]
    return line


class ProgressReporter:
    """Counts completed simulations and draws the status line.

    `enabled=None` means auto: draw only when the stream is a terminal, so
    redirected output and CI logs do not fill up with carriage returns.
    """

    def __init__(
        self,
        total,
        stream=None,
        enabled=None,
        min_interval=MIN_REDRAW_SECONDS,
        clock=time.monotonic,
        bar_width=BAR_WIDTH,
    ):
        self.total = max(0, int(total))
        self.stream = sys.stderr if stream is None else stream
        self.enabled = _is_interactive(self.stream) if enabled is None else bool(enabled)
        self.min_interval = min_interval
        self.bar_width = bar_width
        self._clock = clock
        self._ascii_only = not _supports_block_glyphs(self.stream)
        self.completed = 0
        self._started_at = None
        self._final_elapsed = None
        self._last_draw = None
        self._last_width = 0
        self._drew_anything = False

    @property
    def elapsed(self) -> float:
        """Seconds since start; frozen at the final total once finished."""
        if self._started_at is None:
            return self._final_elapsed or 0.0
        return max(0.0, self._clock() - self._started_at)

    @property
    def rate(self):
        """Average simulations per second, or None before anything finishes."""
        elapsed = self.elapsed
        if elapsed <= 0 or self.completed <= 0:
            return None
        return self.completed / elapsed

    def start(self):
        self._started_at = self._clock()
        self._final_elapsed = None
        self._draw(force=True)
        return self

    def advance(self, count=1):
        self.completed += count
        self._draw()

    def finish(self):
        """Draw the final state and release the line."""
        if self._started_at is None:
            return
        self._draw(force=True)
        if self._drew_anything:
            self.stream.write("\n")
            self._flush()
        self._final_elapsed = self.elapsed
        self._started_at = None

    def track(self, iterable):
        """Yield every item of `iterable`, counting each one as it passes."""
        self.start()
        try:
            for item in iterable:
                self.advance()
                yield item
        finally:
            self.finish()

    def _draw(self, force=False):
        if not self.enabled or self.total <= 0:
            return
        now = self._clock()
        if not force and self._last_draw is not None and now - self._last_draw < self.min_interval:
            return
        self._last_draw = now
        width = _terminal_width(self.stream)
        line = render_line(
            self.completed,
            self.total,
            self.elapsed,
            width=width - 1 if width else None,
            bar_width=self.bar_width,
            ascii_only=self._ascii_only,
        )
        # Pad to the previous line's width so a shorter line cannot leave
        # stale characters behind from the one it overwrites.
        self.stream.write("\r" + line.ljust(self._last_width))
        self._last_width = len(line)
        self._flush()
        self._drew_anything = True

    def _flush(self):
        flush = getattr(self.stream, "flush", None)
        if flush is not None:
            flush()


def _is_interactive(stream) -> bool:
    isatty = getattr(stream, "isatty", None)
    try:
        return bool(isatty and isatty())
    except ValueError:  # closed stream
        return False


def _supports_block_glyphs(stream) -> bool:
    encoding = getattr(stream, "encoding", None) or ""
    try:
        "".join(_BLOCKS).encode(encoding or "utf-8")
    except (LookupError, UnicodeEncodeError):
        return False
    return True


def _terminal_width(stream):
    if not _is_interactive(stream):
        return None
    try:
        return shutil.get_terminal_size().columns
    except (OSError, ValueError):
        return None

File: farm-c/src/progress.c
#include "progress.h"

#include <math.h>
#include <string.h>
#include <time.h>

double progress_now_seconds(void) {
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return (double)ts.tv_sec + (double)ts.tv_nsec / 1e9;
}

void progress_format_duration(double seconds, char *buf, size_t buf_size) {
    if (isnan(seconds) || isinf(seconds)) {
        snprintf(buf, buf_size, "--:--");
        return;
    }
    long total = (long)llround(seconds);
    if (total < 0) total = 0;
    long hours = total / 3600;
    long remainder = total % 3600;
    long minutes = remainder / 60;
    long secs = remainder % 60;
    if (hours > 0) {
        snprintf(buf, buf_size, "%ld:%02ld:%02ld", hours, minutes, secs);
    } else {
        snprintf(buf, buf_size, "%02ld:%02ld", minutes, secs);
    }
}

/* Right-justifies a non-negative integer's decimal digits with ',' every
 * three digits (e.g. 12345 -> "12,345"). `buf_size` must be large enough
 * for the full grouped text plus a NUL; truncates (still NUL-terminated)
 * rather than overflow if not. */
static void format_grouped(long value, char *buf, size_t buf_size) {
    if (value < 0) value = 0; /* rates/counts here are never negative */
    char digits[32];
    snprintf(digits, sizeof(digits), "%ld", value);
    size_t ndigits = strlen(digits);

    size_t out = 0;
    for (size_t i = 0; i < ndigits; i++) {
        if (i > 0 && (ndigits - i) % 3 == 0 && out + 1 < buf_size) buf[out++] = ',';
        if (out + 1 < buf_size) buf[out++] = digits[i];
    }
    buf[out < buf_size ? out : buf_size - 1] = '\0';
}

void progress_format_rate(double rate, char *buf, size_t buf_size) {
    if (isnan(rate) || isinf(rate) || rate < 0.0) {
        snprintf(buf, buf_size, "--");
        return;
    }
    if (rate >= 100.0) {
        format_grouped((long)llround(rate), buf, buf_size);
        return;
    }
    if (rate >= 10.0) {
        snprintf(buf, buf_size, "%.1f", rate);
        return;
    }
    snprintf(buf, buf_size, "%.2f", rate);
}

void progress_render_line(size_t completed, size_t total, double elapsed, char *buf,
                          size_t buf_size, int width, int bar_width, bool ascii_only) {
    double fraction;
    if (total == 0) {
        fraction = 1.0;
    } else {
        fraction = (double)completed / (double)total;
        if (fraction < 0.0) fraction = 0.0;
        if (fraction > 1.0) fraction = 1.0;
    }

    bool has_rate = elapsed > 0.0 && completed > 0;
    double rate = has_rate ? (double)completed / elapsed : NAN;
    double remaining = has_rate ? ((double)total - (double)completed) / rate : NAN;
    if (completed >= total) remaining = 0.0;

    if (bar_width < 1) bar_width = 1;
    /* `bar` below is sized for PROGRESS_BAR_WIDTH cells; every caller in
     * this codebase passes that constant, but clamp defensively rather
     * than trust it. */
    if (bar_width > PROGRESS_BAR_WIDTH) bar_width = PROGRESS_BAR_WIDTH;
    int filled_cells = (int)(fraction * (double)bar_width);
    if (filled_cells > bar_width) filled_cells = bar_width;
    if (filled_cells < 0) filled_cells = 0;
    int empty_cells = bar_width - filled_cells;

    /* U+2588 FULL BLOCK / U+2591 LIGHT SHADE, matching runner/progress.py's
     * _BLOCKS -- same glyphs, so a terminal that renders one renders both. */
    const char *filled_glyph = ascii_only ? "#" : "\xE2\x96\x88";
    const char *empty_glyph = ascii_only ? "-" : "\xE2\x96\x91";

    char bar[PROGRESS_BAR_WIDTH * 3 + 1];
    size_t bar_len = 0;
    for (int i = 0; i < filled_cells && bar_len + 4 <= sizeof(bar); i++) {
        size_t glyph_len = strlen(filled_glyph);
        memcpy(bar + bar_len, filled_glyph, glyph_len);
        bar_len += glyph_len;
    }
    for (int i = 0; i < empty_cells && bar_len + 4 <= sizeof(bar); i++) {
        size_t glyph_len = strlen(empty_glyph);
        memcpy(bar + bar_len, empty_glyph, glyph_len);
        bar_len += glyph_len;
    }
    bar[bar_len] = '\0';

    /* Floored, not rounded: 9,999 of 10,000 runs must not read as 100%. */
    double percent = floor(fraction * 1000.0) / 10.0;

    char total_text[32];
    format_grouped((long)total, total_text, sizeof(total_text));
    char completed_text[32];
    format_grouped((long)completed, completed_text, sizeof(completed_text));

    char rate_text[16];
    progress_format_rate(rate, rate_text, sizeof(rate_text));
    char elapsed_text[16];
    progress_format_duration(elapsed, elapsed_text, sizeof(elapsed_text));
    char remaining_text[16];
    progress_format_duration(remaining, remaining_text, sizeof(remaining_text));

    size_t total_width = strlen(total_text);
    size_t completed_width = strlen(completed_text);
    int pad = (int)(total_width > completed_width ? total_width - completed_width : 0);

    int written = snprintf(buf, buf_size, "[%s] %5.1f%% | %*s/%s | %s sim/s | %s elapsed | %s left",
                           bar, percent, pad + (int)completed_width, completed_text, total_text,
                           rate_text, elapsed_text, remaining_text);

    if (width > 0 && written > width && (size_t)written < buf_size) buf[width] = '\0';
}

static void draw(ProgressReporter *p, bool force) {
    if (!p->enabled || p->total == 0) return;
    double now = progress_now_seconds();
    if (!force && p->has_last_draw && (now - p->last_draw_at) < p->min_interval_seconds) return;
    p->last_draw_at = now;
    p->has_last_draw = true;

    char line[PROGRESS_LINE_MAX];
    progress_render_line(p->completed, p->total, progress_elapsed_seconds(p), line, sizeof(line),
                         0, PROGRESS_BAR_WIDTH, p->ascii_only);
    size_t len = strlen(line);

    fputc('\r', p->stream);
    fputs(line, p->stream);
    /* Pad to the previous line's width so a shorter line cannot leave stale
     * characters behind from the one it overwrites. */
    for (size_t i = len; i < p->last_width; i++) fputc(' ', p->stream);
    p->last_width = len;
    fflush(p->stream);
    p->drew_anything = true;
}

void progress_init(ProgressReporter *p, size_t total, FILE *stream, bool enabled) {
    memset(p, 0, sizeof(*p));
    p->total = total;
    p->stream = stream != NULL ? stream : stderr;
    p->enabled = enabled;
    p->min_interval_seconds = 0.1;
    p->ascii_only = false;
}

void progress_start(ProgressReporter *p) {
    p->started_at = progress_now_seconds();
    p->running = true;
    p->final_elapsed = 0.0;
    p->has_last_draw = false;
    p->last_width = 0;
    p->drew_anything = false;
    p->completed = 0;
    draw(p, true);
}

double progress_elapsed_seconds(const ProgressReporter *p) {
    if (!p->running) return p->final_elapsed;
    double elapsed = progress_now_seconds() - p->started_at;
    return elapsed < 0.0 ? 0.0 : elapsed;
}

void progress_advance(ProgressReporter *p, size_t count) {
    p->completed += count;
    draw(p, false);
}

void progress_finish(ProgressReporter *p) {
    if (!p->running) return;
    draw(p, true);
    if (p->drew_anything) {
        fputc('\n', p->stream);
        fflush(p->stream);
    }
    p->final_elapsed = progress_elapsed_seconds(p);
    p->running = false;
}
